feat: add initial implementation of Saleor GraphQL client

Ang-m4 opened the pull request:
## Description
This PR introduces a new GraphQL client as part of the tutor-saleor plugin. Its primary function is to import courses into saleor-core through its API. The client provides all the necessary methods to create multiple courses based on the Open edX CourseOverview model.

## Changes
- Added a new `mutations.py` file that contains all mutations required for the import operation.
- Added a new `queries.py` file that includes all queries needed for the import operation.
- Added a new `utils.py` file that provides utilities for parsing attributes and generating mutation/query strings.
- Added a new `client.py` file that manages client initialization and error handling.

## Steps
Before importing a course into the Saleor API—given its unique schema—it is essential to perform the following steps:

1. **Create Attributes:** Saleor employs a completely new attribute schema. Therefore, it is necessary to identify the course overview attributes and create a corresponding attribute instance for each.
2. **Create the Product Type:** Attributes can be grouped into a "productType" model for later assignment to products. In our case, this product type could be named "course" or "online-course."
3. **Create the Course Product:** Once the product type and its associated attributes are successfully created, the course import process can proceed. It is important to have the product type ID and the input types of the attributes correctly mapped.

## How to test?

1. Clone this branch into the `/mnt` directory of your Tutor environment.
2. Ensure the following environment variables are set:
   ```bash
   SALEOR_API_URL
   SALEOR_API_TOKEN
   ```
3. Access your LMS container with:
   ```bash
   docker exec -it tutor_dev-lms-1 bash
   ```
   Then install the plugin:
   ```bash
   pip install -e /mnt/platform-plugin-saleor/
   ```
4. Restart the LMS container and open a Python shell:
   ```bash
   ./manage.py lms shell
   ```
5. Inside the shell, import the client and run the following:
   ```python
   from platform_plugin_saleor.saleor_client.client import SaleorApiClient

   client = SaleorApiClient()
   client.create_base_product_attributes()
   client.create_course_product_type()
   ```
6. Once the product type and attributes have been created, import the course overview and test the course load into Saleor:
   ```python
   from openedx.core.djangoapps.content.course_overviews.models import CourseOverview

   courses = CourseOverview.objects.all()
   for course in courses:
       client.create_course_product(course)
   ```
7. Finally, check the Saleor dashboard to confirm that the courses were successfully created.

![product-type-created](https://github.com/user-attachments/assets/2d0d15ba-c63e-44db-8a40-2291eba80787)

![attributes-created](https://github.com/user-attachments/assets/dc39687c-25dd-4293-ba86-b49e18961c8c)

![products-created](https://github.com/user-attachments/assets/d9354695-f757-473d-9e33-712d0c0883e3)

![attributes-filledpng](https://github.com/user-attachments/assets/a103f0ac-5478-45ff-8f62-f46eeb4dd923)

Ang-m4 committed:
feat: add initial implementation of Saleor GraphQL client

Ang-m4 committed:
fix: remove unnecessary ci docs job

johanseto commented on platform_plugin_saleor/saleor_client/client.py:
```diff
@@ -0,0 +1,129 @@
+        self.token = token or settings.SALEOR_API_TOKEN
+
+        try:
+            transport = AIOHTTPTransport(
```
AI is thinked to be used with async?

johanseto commented on platform_plugin_saleor/saleor_client/client.py:
```diff
@@ -0,0 +1,129 @@
+            return response
+        except Exception as e:  # pylint: disable=broad-exception-caught
+            logger.error(f"Failed to create course product: {str(e)}")
+            return {}
```
Its is possible to create a general execute like this example https://github.com/mirumee/saleor-app-framework-python/blob/main/src/saleor_app/saleor/client.py#L37, there we could add the attributes and other things... Also the execute could be reused for the other methods.

this is something to improve, if not we could move on

Ang-m4 committed:
fix: enhance saleor client

- improve the mutations and query definition
- define a new graphql exceptions to better error handling
- simplify mutations creation logic laveragin graphql variables

Ang-m4 commented:
Hi @johanseto! 
I took a look and totally agree. Having an `execute` method really makes the usage cleaner and more intuitive. I went ahead and refactored the entire client logic to follow that approach, and also added some minor error handling improvements.

Additionally, I was previously building the queries dynamically using string interpolation. Now, by leveraging the client's variables properly, everything should work as expected and be much more robust.

Please let me know what you think!

andrey-canon commented on platform_plugin_saleor/saleor_client/client.py:
```diff
@@ -0,0 +1,70 @@
+"""Saleor API client for creating course products and attributes."""
+
+
```
is this necessary ?

andrey-canon commented on platform_plugin_saleor/saleor_client/client.py:
```diff
@@ -0,0 +1,70 @@
+        self.token = token or settings.SALEOR_API_TOKEN
+
+        try:
+            transport = AIOHTTPTransport(
```
the standard request library is not compatible with graphql, is it ?

Ang-m4 replied:
Since GraphQL operates over HTTP, we could achieve the same behavior using the standard requests library.

andrey-canon commented on platform_plugin_saleor/saleor_client/client.py:
```diff
@@ -0,0 +1,70 @@
+            token: Saleor API token, defaults to settings.SALEOR_API_TOKEN
+            timeout: Request timeout in seconds.
+        """
+        self.base_url = base_url or settings.SALEOR_API_URL
```
could you remove the dependency  from django settings?, the current implementation is ok but if we have to move this to an independent package this will require implementation changes

andrey-canon commented on platform_plugin_saleor/saleor_client/client.py:
```diff
@@ -0,0 +1,70 @@
+            timeout: Request timeout in seconds.
+        """
+        self.base_url = base_url or settings.SALEOR_API_URL
+        self.token = token or settings.SALEOR_API_TOKEN
```
same comment as above

andrey-canon commented on platform_plugin_saleor/saleor_client/client.py:
```diff
@@ -0,0 +1,70 @@
+            )
+        except Exception as e:
+            logger.error(f"Failed to initialize Saleor API client: {str(e)}")
+            raise
```
This could be considered redundant in a sentry environment,  the logger error sends a sentry notification then the raise will send another sentry notification, could you use a custom exception or remove the try-except since you are not handling any exception

andrey-canon commented on platform_plugin_saleor/saleor_client/client.py:
```diff
@@ -0,0 +1,70 @@
+                transport=transport,
+                fetch_schema_from_transport=False,
+            )
+        except Exception as e:
```
it's better if you identify the the possible exception that can be raised, otherwise the try-except doesn't make a lot of sense since you raise the original exception

andrey-canon commented on platform_plugin_saleor/saleor_client/utils.py:
```diff
@@ -0,0 +1,208 @@
+    if isinstance(response_data, dict):
+        if "errors" in response_data and response_data["errors"]:
+            return response_data["errors"]
+        for _, value in response_data.items():
```
```suggestion
        for value in response_data.values():
```
ref https://www.w3schools.com/python/ref_dictionary_values.asp

andrey-canon commented on platform_plugin_saleor/saleor_client/utils.py:
```diff
@@ -0,0 +1,208 @@
+        if "errors" in response_data and response_data["errors"]:
+            return response_data["errors"]
+        for _, value in response_data.items():
+            errors = find_errors(value)
+            if errors:
```
```suggestion
            if errors := find_errors(value):
```

andrey-canon commented on platform_plugin_saleor/saleor_client/utils.py:
```diff
@@ -0,0 +1,208 @@
+    elif isinstance(response_data, list):
+        for item in response_data:
+            errors = find_errors(item)
+            if errors:
```
```suggestion
            if errors := find_errors(item):
```

andrey-canon commented on platform_plugin_saleor/saleor_client/utils.py:
```diff
@@ -0,0 +1,208 @@
+
+
+def generate_saleor_product_attribute_data(
+    model_cls, model_field: str, product_attrb_name: str
```
```suggestion
    model_cls, model_field: str, product_attr_name: str
```
?

andrey-canon commented on platform_plugin_saleor/saleor_client/utils.py:
```diff
@@ -0,0 +1,208 @@
+
+    for edge in edges:
+        node = edge.get("node", {})
+        if node:
```
```suggestion
        if node := edge.get("node", {}):
```

andrey-canon commented on platform_plugin_saleor/saleor_client/utils.py:
```diff
@@ -0,0 +1,208 @@
+                'product_attribute': 'Self Paced',
+            },
+            {
+                'model_attribute': 'mobile_available',
```
do we need this ?

andrey-canon commented on platform_plugin_saleor/saleor_client/utils.py:
```diff
@@ -0,0 +1,208 @@
+                'product_attribute': 'Mobile Available',
+            },
+            {
+                'model_attribute': 'cert_name_long',
```
same comment as above

andrey-canon commented on platform_plugin_saleor/saleor_client/utils.py:
```diff
@@ -0,0 +1,208 @@
+                'product_attribute': 'Certificate Name',
+            },
+            {
+                'model_attribute': 'lowest_passing_grade',
```
same comment as above

andrey-canon commented on platform_plugin_saleor/saleor_client/utils.py:
```diff
@@ -0,0 +1,208 @@
+                'product_attribute': 'Lowest Passing Grade',
+            },
+            {
+                'model_attribute': 'invitation_only',
```
same comment as above

andrey-canon commented on platform_plugin_saleor/saleor_client/utils.py:
```diff
@@ -0,0 +1,208 @@
+            },
+            {
+                'model_attribute': 'max_student_enrollments_allowed',
+                'product_attribute': 'Max Student Enrollments Allowed',
```
same comment as above

andrey-canon commented on platform_plugin_saleor/saleor_client/utils.py:
```diff
@@ -0,0 +1,208 @@
+                'product_attribute': 'Language'
+            },
+            {
+                'model_attribute': 'entrance_exam_minimum_score_pct',
```
same comment as above

Ang-m4 committed:
feat: define course creation logic at client

Ang-m4 committed:
refactor: simplify SaleorApiClient initialization and improve attribute naming
